lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java

```diff
@@ -152,6 +154,51 @@ void renameFile(Location source, Location target)
     FileIterator listFiles(Location location)
             throws IOException;
 
+    default FileIterator listFilesNonRecursive(Location location)
+            throws IOException
+    {
+        FileIterator recursiveIterator = listFiles(location);
+        return new FileIterator() {
+
+            @Nullable
+            private Optional<FileEntry> next;
```

Comment:
define what `null` means and what `empty` means.

`com.google.common.collect.AbstractIterator` solves this with explicit state enum.